src/contract.rs

```diff
@@ -993,9 +993,7 @@ fn use_allowance<S: Storage>(
 ) -> StdResult<()> {
     let mut allowance = read_allowance(storage, owner, spender)?;
 
-    if allowance.expiration.map(|ex| ex < env.block.time) == Some(true) && allowance.amount != 0 {
-        allowance.amount = 0;
-        write_allowance(storage, owner, spender, allowance)?;
+    if allowance.is_expired_at(&env.block) {
         return Err(insufficient_allowance(0, amount));
```

Comment:
Note to reviewer: This code segment was expensive and unnecessary. There's no need to actively set the amount to 0 if the allowance query would make it clear to consumers that the allowance has expired anyway (so the amount isn't important)